### PR TITLE
Fix keyword-based search on /search/_

### LINKED ISSUE #2420: Fix keyword-based search on /search/____
Hi, this is a [first-timers-only issue](https://publiclab.github.io/community-toolbox/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

This will help make searches like https://publiclab.org/search/pumpkin to properly display results like this page:

https://publiclab.org/notes/cfastie/11-22-2013/white-balance-for-all

As described in this issue: https://github.com/publiclab/plots2/issues/2411

Thank you!!

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Update** the file [app/services/search_service.rb](https://github.com/publiclab/plots2/blob/master/app/services/search_service.rb) in the `plots2` repository (press the little pen Icon) and edit the line as shown below. 

[See this page](https://publiclab.github.io/community-toolbox/#r=all) for some help in taking your first steps!

Below is a "diff" showing in red (and a `-`) which lines to remove, and in green (and a `+`) which lines to add:

```diff
@@ -54,17 +54,21 @@ def find_comments(input, limit = 5)
   end
 
   def find_nodes(input, limit = 5)
-    Node.limit(limit)
-      .order('nid DESC')
-      .where('node.status = 1 AND title LIKE ?', '%' + input + '%')
+    if ActiveRecord::Base.connection.adapter_name == 'Mysql2'
+      Node.search(input)
+        .group(:nid)
+        .includes(:node)
+        .references(:node)
+        .limit(limit)
+        .where("node.type": ["note", "page"], "node.status": 1)
+        .order('node.changed DESC')
+    else 
+      Node.limit(limit)
+        .group(:nid)
+        .where(type: ["note", "page"], status: 1)
+        .order(changed: :desc)
+        .where('title LIKE ?', '%' + input + '%')
+    end
  end

-  ## search for node title only
-  ## FIXme with solr
  def find_notes(input, limit = 5)
    Node.limit(limit)
      .order('nid DESC')
      .where('type = "note" AND node.status = 1 AND title LIKE ?', '%' + input + '%')
   end
 
   def find_maps(input, limit = 5)
```

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### 🤔❓ Questions?

Leave a comment below!

### Is someone else already working on this?

We encourage you to link to this issue by mentioning the issue # in your pull request, so we can see if someone's already started on it. **If someone seem stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://publiclab.github.io/community-toolbox/#r=all). Thanks!

(This issue was created by [First-Timers-Bot](https://github.com/hoodiehq/first-timers-bot).)

### PR DESCRIPTION
**Pull request by chitra591**
Fixes #2420

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by chitra591** [comment]
Hey @jywarren ,

Really sorry to trouble you, but the test passed in my local system, and I haven't changed the file searches_controller_test.rb for which the test fails.

Do you want me to do anything?

screen shot 2018-03-01 at 9 36 46 pm

**Comment by jywarren** [comment]
Hmm, I'm puzzling over this one. The change you made does affect searches controller, so there is a relationship, but I can't work out what's gone wrong! 

Do you know how to use a rails console by running `rails console` and then you can try running `SearchService.new.nodes("About")` to see what you get? It should not return an empty array, although that's what this test is telling us it does. 

Maybe that'll be a clue! If not, I'll download your code and try it out and maybe there's something we didn't anticipate going on. 

Thanks for working on this!!!

**Comment by chitra591** [comment]
Thanks @jywarren !

I get one record as a response (Note: I have connected to Sqlite, as suggested in installation guide)

screen shot 2018-03-02 at 8 40 12 am

**Comment by jywarren** [comment]
Ahhhhh well the top section of code runs in MySQL only... That explains the
discrepancy!

I can try to run it on my machine tomorrow, as it's late here. We'll see!

**Comment by chitra591** [comment]
Hey @jywarren,

I switched my db to mysql and dug a little deeper into this issue and found something strange! The issue is only for the string 'About'.

The `SearchService.new.nodes("About") ` returns an empty array, but returns data for all other search. For example `SearchService.new.nodes("Events")` 

search result for about

search result for events


I checked the database too, even though there is a record for About, it returns an emptySet.

screen shot of db

db - search about
db - search - events


Hope this helps!

**Comment by jywarren** [comment]
OMG -- great sleuthing! Amazing! I thought about that a bit and it triggered an odd thought -- the "full text search" we use now is a bit mysterious, and I remembered that it discards what it thinks of as "common" words!

Here are the docs: https://dev.mysql.com/doc/refman/5.7/en/fulltext-stopwords.html

You can see the word "about" near the top!

WOW! What an obscure catch -- and a tough one to pin down! Fantastic work! 

Shall we then change the word we use to search for to something else from our test fixtures? Say, `Organizers`:

https://github.com/publiclab/plots2/blob/2d43c7d908dd8b65d3ca7681cc35a107770e0f20/test/fixtures/nodes.yml#L49-L53

Hooray! 👍 👍 🎉

**Comment by chitra591** [comment]
Thanks!

I think it would be great, if we change the word we use to search for to something else from the test fixtures.
But we don't have Organizers in the development DB. The test would still always fail in local system.

Some of the available nodes are -- Media, Events, Getting-started, Donate, Stats, Licenses, Map etc.
We could use any of the above or add Organizers in the development DB too.

:)

**Comment by jywarren** [comment]
Ah, I think it's ok- the tests run on the test fixture data and not the
development seed data, so although it won't work on the local copy you
have, I think it will if you make the change to the test and push it up to
run tests. Give it a try!

**Comment by chitra591** [comment]
Hey @jywarren !

It is still not working!  :(
This is the Error -- 
`ActionView::Template::Error: undefined method 'tagnames_as_classes' for #<Revision:0x000000000ae9db30>`

As far as I understand,  `tagnames_as_classes` is a method of `Node`, instead of `Revision`. But I don't know why is it calling this method for a Revision's object! 

Need your help.

**Comment by Gauravano** [comment]
This is strange, I am not getting any error after reflecting these changes and running rake test.

**Comment by chitra591** [comment]
Is it?
All the rake tests passed even when you connected to MySQL?

That's strange!

**Comment by chitra591** [comment]
Hi @jywarren,
Can you please take a look at this, when you get time? Let me know if I need to do any work on this.

Thanks!

**Comment by ebarry** [comment]
adding in @bronwen9 to help review -- Yay Librarians! Hooray for the information sciences!

**Comment by jywarren** [comment]
Aha - so the `node.search` method (for better or worse) actually returns `revisions` not `nodes`. So it's not a completely simple swap-in, but you can add `revision.parent.tagnames_as_classes` to resolve this. 

Sorry I was traveling for a few days! Hope this helps!

**Comment by chitra591** [comment]
Sorry for the late response.. I had missed your message.

And that's correct, `find_nodes` gave an array of `Revisions` for Mysql DB connection where as `Node` for other DB connections. 
So I mapped the array to `Revisions` to `Node`, using the `parent` method.

Now all the tests have passed!! and I think you could merge it, if it looks good.

Thanks for the help @jywarren :)

File: app/services/search_service.rb [hunk 1]
```diff
@@ -54,13 +54,23 @@ def find_comments(input, limit = 5)
...
-      .order('nid DESC')
-      .where('node.status = 1 AND title LIKE ?', '%' + input + '%')
+    if ActiveRecord::Base.connection.adapter_name == 'Mysql2'
+      Node.search(input)
+        .group(:nid)
+        .includes(:node)
+        .references(:node)
+        .limit(limit)
+        .where("node.type": ["note", "page"], "node.status": 1)
+        .order('node.changed DESC').map{ |x| x.parent}
```
Comment (jywarren) [inline app/services/search_service.rb]: OK, this is pretty good. But the problem is that then the `.map()` function will run once for each result -- not very efficient! Instead we could do:

```ruby
nids = Node.search ...
```

up on line 58, and then at the end:


```ruby
        .order('node.changed DESC')
        .collect(&:nid)
Node.find nids
```

That way we get them all in just one single query. It's not super pretty, but it's efficient!

**Comment by jywarren** [comment]
Amazingggggg!!!!!! @ebarry will be sooooooo happy! 👍 👍 👍 🎉

**Comment by jywarren** [comment]
So grateful!

**Comment by ebarry** [comment]
WOWOWOWOWOWOW!!!!! THANK YOU SO MUCHNESS!!!!!!!! HAPPYTASTIC!!!!!!!

**Pull request merged**